content/posts/2018-10-21--complex-command-handler-in-javascript.md

```diff
@@ -0,0 +1,352 @@
+---
+title: Complex command handler in JavaScript
+subTitle: Comparison of using Promise and async/await
+cover: /img/js.jpg
+tags: ["JavaScript", "Promise", "async", "await"]
+postAuthor: Mariusz Bąk (malef)
+---
+Recently I needed to carry out some complicated process in my Node.js side project. This process consisted of multiple steps and while some of them could be run in parallel some other required to be run in sequence. Here I will show the basic solution I came up with as it's quite short and demonstrates the way of using both `Promise`s and `async`/`await`.
+
+I wanted to follow Command design pattern (https://en.wikipedia.org/wiki/Command_pattern) where Command object representing a single operation is provided with all the data required for carrying it out. Later such Command object is provided to a dedicated CommandHandler that would actually perform the task at hand, possibly making use of some external dependencies.
+
+In other words, I needed a piece of code that will execute some composite command that I would define upfront. This composite command would describe the whole process and would be composed of other composite commands (that would be handled recursively) or simple commands (that would have their dedicated command handlers provided as services). Let's go through a process of building such a service.
+
+First let's prepare some classes that will represent our commands. We need two classes for it: one for the atomic command that will be provided to a dedicated CommandHandler, and one composite command for composing multiple commands (either atomic or composite) into a set that can be handled either in parallel or in sequence. Let's call them SimpleCommand and CompositeCommand --- they will be quite basic.
+
+For the sake of simplicity our atomic command will not describe any real operations --- we will provide it only with an identifier so that we can track if everything works fine:
```

Comment:
```suggestion
For the sake of simplicity, our atomic command will not describe any real operations --- we will provide it only with an identifier so that we can track if everything works fine:
```